packages/fxa-auth-server/scripts/cancel-subscriptions-to-plan/cancel-subscriptions-to-plan.ts

```diff
@@ -0,0 +1,261 @@
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const subscriptionCollection = `${collectionPrefix}subscriptions`;
+
+    const subscriptionSnap = await this.firestore
+      .collectionGroup(subscriptionCollection)
+      .where('plan.id', '==', this.priceId)
+      .orderBy('id')
+      .startAfter(startAfter)
+      .limit(this.batchSize)
+      .get();
+
+    const subscriptions = subscriptionSnap.docs.map(
+      (doc) => doc.data() as FirestoreSubscription
+    );
+
+    return subscriptions;
+  }
+
+  /**
+   * Attempts to cancel a firestore subscription
+   * @param firestoreSubscription The subscription to cancel
+   */
+  async processSubscription(firestoreSubscription: FirestoreSubscription) {
+    const { id: subscriptionId, customer: customerId } = firestoreSubscription;
+
+    try {
+      const customer = await this.fetchCustomer(customerId);
+      if (!customer?.subscriptions?.data) {
+        console.error(`Customer not found: ${customerId}`);
+        return;
+      }
+
+      const account = await this.database.account(customer.metadata.userid);
+      if (!account) {
+        console.error(`Account not found: ${customer.metadata.userid}`);
+        return;
+      }
+
+      const isExcluded = this.isCustomerExcluded(customer.subscriptions.data);
+
+      if (!this.dryRun && !isExcluded) {
+        await this.cancelSubscription(firestoreSubscription);
+      }
+
+      const report = this.buildReport(customer, account, isExcluded);
+
+      await this.writeReport(report);
+
+      console.log(subscriptionId);
+    } catch (e) {
+      console.error(subscriptionId, e);
+    }
+  }
+
+  /**
+   * Retrieves a customer record directly from Stripe
+   * @param customerId The Stripe customer ID of the customer to fetch
+   * @returns The customer record for the customerId provided, or null if not found or deleted
+   */
+  async fetchCustomer(customerId: string) {
+    const customer = await this.enqueueRequest(() =>
+      this.stripe.customers.retrieve(customerId, {
+        expand: ['subscriptions'],
+      })
+    );
+
+    if (customer.deleted) return null;
+
+    return customer;
+  }
+
+  /**
+   * Cancel subscription and refund customer for latest bill
+   * @param subscription The subscription to cancel
+   */
+  async cancelSubscription(subscription: Stripe.Subscription) {
+    await this.enqueueRequest(() =>
+      this.stripe.subscriptions.cancel(subscription.id, {
+        prorate: this.prorate,
+      })
+    );
+
+    if (!this.refund) return;
+
+    if (!subscription.latest_invoice) {
+      console.log(`No latest invoice for ${subscription.id}`);
+      return;
+    }
+
+    const latestInvoiceId =
+      typeof subscription.latest_invoice === 'string'
+        ? subscription.latest_invoice
+        : subscription.latest_invoice.id;
+
+    const invoice = await this.enqueueRequest(() =>
+      this.stripe.invoices.retrieve(latestInvoiceId)
```

Comment:
Is this call to retrieve an invoice necessary if `latest_invoice` is not a string (i.e. its an invoice)?
Not sure if saving an extra stripe call on this script matters, but pointing this out in case it does.

Reply:
It's not technically necessary in the case that latest_invoice is not a string.

I do this because this object is from Firestore so I'm using a fair amount of caution here and re-fetching in case the last invoice in Firestore is stale, especially right before we issue a credit directly to the user's card.